File: examples/dekita/.claude/hooks/flow_progress_tracker.py
#!/usr/bin/env python3
"""フローステップの完了を追跡する。

Why:
    開発ワークフローの各ステップ（worktree作成、コミット、PR作成等）が
    実際に実行されたことを記録する必要がある。コマンドパターンマッチで
    自動的にステップ完了を検出し、flow-effect-verifierが判断に使用する。

What:
    - Bashコマンドの実行を監視
    - flow_definitions.pyのパターンとマッチング
    - マッチした場合はステップを完了としてマーク
    - ステップ順序の妥当性を検証（警告のみ）

State:
    writes: .claude/state/flow-progress.jsonl

Remarks:
    - 終了コードが0のコマンドのみ対象
    - 順序違反は警告するが、ステップは完了としてマーク（回復可能）
"""

import json

from common import complete_flow_step, get_incomplete_flows
from flow_definitions import get_flow_definition, validate_step_order
from lib.execution import log_hook_execution
from lib.hook_input import get_tool_result
from lib.session import parse_hook_input


def get_active_flows() -> list[dict]:
    """Get active (incomplete) flows in current session with flow definitions.

    Uses get_incomplete_flows() from common.py and adds flow_definition objects
    for context-aware pattern matching.

    Returns:
        List of flow entries with their context, completed steps, and flow definitions.
    """
    incomplete = get_incomplete_flows()

    active: list[dict] = []
    for flow in incomplete:
        flow_id = flow.get("flow_id")
        flow_def = get_flow_definition(flow_id)
        if not flow_def:
            continue

        active.append(
            {
                "flow_instance_id": flow.get("flow_instance_id"),
                "flow_id": flow_id,
                "flow_definition": flow_def,
                "context": flow.get("context", {}),
                "pending_steps": flow.get("pending_steps", []),
                "completed_steps": flow.get("completed_steps", []),
            }
        )

    return active


def main() -> None:
    """Main entry point for the hook."""
    input_data = parse_hook_input()

    # Only process Bash tool
    tool_name = input_data.get("tool_name")
    if tool_name != "Bash":
        return

    # Get the command that was executed
    tool_input = input_data.get("tool_input", {})
    command = tool_input.get("command", "")
    if not command:
        return

    # Only mark steps complete if command succeeded (exit_code == 0)
    tool_result = get_tool_result(input_data) or {}
    exit_code = tool_result.get("exit_code", 0)
    if exit_code != 0:
        return  # Don't mark steps complete for failed commands

    # Get active flows
    active_flows = get_active_flows()
    if not active_flows:
        return

    # Check each active flow for matching steps (with context-aware matching)
    matched_steps: list[tuple[str, str, str]] = []  # (instance_id, step_id, step_name)
    order_warnings: list[str] = []

    for flow in active_flows:
        instance_id = flow["flow_instance_id"]
        flow_id = flow["flow_id"]
        flow_def = flow["flow_definition"]
        context = flow["context"]
        completed = flow["completed_steps"]

        for step_id in flow["pending_steps"]:
            # Use context-aware pattern matching from flow_definitions.py
            if flow_def.matches_step(step_id, command, context):
                # Validate step order
                is_valid, error_msg = validate_step_order(flow_id, completed, step_id)

                if not is_valid:
                    order_warnings.append(f"順序警告: {error_msg}")
                    # Still mark the step as complete, but warn about order
                    # This allows recovery from out-of-order execution

                step = flow_def.get_step(step_id)
                step_name = step.name if step else step_id
                matched_steps.append((instance_id, step_id, step_name))
                complete_flow_step(instance_id, step_id, flow_id)
                # Keep local completed list in sync for subsequent validations
                completed.append(step_id)

    # Output notification if steps were completed
    if matched_steps or order_warnings:
        messages = []
        if matched_steps:
            step_names = [name for _, _, name in matched_steps]
            messages.append(f"ステップ完了: {', '.join(step_names)}")
            # Log step completions
            log_hook_execution(
                "flow-progress-tracker",
                "approve",
                f"Steps completed: {', '.join(step_names)}",
                {
                    "matched_steps": [
                        {"instance_id": inst, "step_id": sid, "step_name": name}
                        for inst, sid, name in matched_steps
                    ]
                },
            )
        messages.extend(order_warnings)

        result = {
            "continue": True,
            "systemMessage": "[flow-progress-tracker] " + " | ".join(messages),
        }
        print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/flow_progress_tracker.ts
#!/usr/bin/env bun
/**
 * フローステップの完了を追跡する。
 *
 * Why:
 *   開発ワークフローの各ステップ（worktree作成、コミット、PR作成等）が
 *   実際に実行されたことを記録する必要がある。コマンドパターンマッチで
 *   自動的にステップ完了を検出し、flow-effect-verifierが判断に使用する。
 *
 * What:
 *   - Bashコマンドの実行を監視
 *   - flow_definitions.tsのパターンとマッチング（コンテキスト対応）
 *   - マッチした場合はステップを完了としてマーク
 *   - ステップ順序の妥当性を検証（警告のみ）
 *
 * State:
 *   writes: .claude/logs/flow/flow-progress-{session_id}.jsonl
 *
 * Remarks:
 *   - 終了コードが0のコマンドのみ対象
 *   - 順序違反は警告するが、ステップは完了としてマーク（回復可能）
 *
 * Changelog:
 *   - silenvx/dekita#3051: TypeScriptに移行
 *   - silenvx/dekita#3157: flow_definitions.ts統合、順序検証追加
 */

import { basename } from "node:path";
import { FLOW_LOG_DIR } from "../lib/common";
import { completeFlowStep, getIncompleteFlows } from "../lib/flow";
import { type FlowDefinition, getFlowDefinition, validateStepOrder } from "../lib/flow_definitions";
import { getExitCode, getToolResult } from "../lib/input_context";
import { logHookExecution } from "../lib/logging";
import { createContext, parseHookInput } from "../lib/session";

const HOOK_NAME = "flow-progress-tracker";

// =============================================================================
// Types
// =============================================================================

interface ActiveFlow {
  flow_instance_id: string;
  flow_id: string;
  flow_definition: FlowDefinition;
  context: Record<string, unknown>;
  pending_steps: string[];
  completed_steps: string[];
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Get active (incomplete) flows in current session with flow definitions.
 *
 * Uses getIncompleteFlows() from flow.ts and adds flow_definition objects
 * for context-aware pattern matching.
 *
 * @param sessionId - Session ID for file isolation.
 * @returns List of flow entries with their context, completed steps, and flow definitions.
 */
async function getActiveFlows(sessionId?: string): Promise<ActiveFlow[]> {
  const incomplete = await getIncompleteFlows(FLOW_LOG_DIR, sessionId);

  const active: ActiveFlow[] = [];
  for (const flow of incomplete) {
    const flowId = flow.flow_id;
    const flowDef = getFlowDefinition(flowId);
    if (!flowDef) {
      continue;
    }

    active.push({
      flow_instance_id: flow.flow_instance_id,
      flow_id: flowId,
      flow_definition: flowDef,
      context: flow.context,
      pending_steps: flow.pending_steps,
      completed_steps: [...flow.completed_steps],
    });
  }

  return active;
}

// =============================================================================
// Main
// =============================================================================

async function main(): Promise<void> {
  const result: { continue: boolean; systemMessage?: string } = {
    continue: true,
  };

  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    const ctx = createContext(inputData);

    sessionId = ctx.sessionId ?? undefined;
    // Sanitize sessionId early to prevent path traversal in all uses
    const safeSessionId = sessionId ? basename(sessionId) : undefined;

    // Only process Bash tool
    const toolName = inputData.tool_name;
    if (toolName !== "Bash") {
      console.log(JSON.stringify(result));
      return;
    }

    // Get the command that was executed
    const toolInput = inputData.tool_input ?? {};
    const command = typeof toolInput.command === "string" ? toolInput.command : "";
    if (!command) {
      console.log(JSON.stringify(result));
      return;
    }

    // Only process if command succeeded (exit_code == 0)
    // getExitCode defaults to 0 for missing exit_code (success), but we want to
    // treat missing as failure (-1) to be conservative about tracking
    const toolResult = getToolResult(inputData as Record<string, unknown>);
    const exitCode = getExitCode(toolResult, -1);
    if (exitCode !== 0) {
      console.log(JSON.stringify(result));
      return;
    }

    // Get active (incomplete) flows with flow definitions
    const activeFlows = await getActiveFlows(safeSessionId);

    if (activeFlows.length === 0) {
      console.log(JSON.stringify(result));
      return;
    }

    // Check each active flow for matching steps (with context-aware matching)
    const matchedSteps: Array<{ instanceId: string; stepId: string; stepName: string }> = [];
    const orderWarnings: string[] = [];

    for (const flow of activeFlows) {
      const instanceId = flow.flow_instance_id;
      const flowId = flow.flow_id;
      const flowDef = flow.flow_definition;
      const context = flow.context;
      const completed = flow.completed_steps;

      for (const stepId of flow.pending_steps) {
        // Use context-aware pattern matching from flow_definitions.ts
        if (flowDef.matchesStep(stepId, command, context)) {
          // Validate step order
          const [valid, errorMessage] = validateStepOrder(flowId, completed, stepId);

          if (!valid && errorMessage) {
            orderWarnings.push(`順序警告: ${errorMessage}`);
            // Still mark the step as complete, but warn about order
            // This allows recovery from out-of-order execution
          }

          const step = flowDef.getStep(stepId);
          const stepName = step?.name ?? stepId;
          matchedSteps.push({ instanceId, stepId, stepName });

          await completeFlowStep(FLOW_LOG_DIR, instanceId, stepId, flowId, safeSessionId);

          // Keep local completed list in sync for subsequent validations
          completed.push(stepId);
        }
      }
    }

    // Output notification if steps were completed
    if (matchedSteps.length > 0 || orderWarnings.length > 0) {
      const messages: string[] = [];

      if (matchedSteps.length > 0) {
        const stepNames = matchedSteps.map((s) => s.stepName);
        messages.push(`ステップ完了: ${stepNames.join(", ")}`);

        // Log step completions
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `Steps completed: ${stepNames.join(", ")}`,
          {
            matched_steps: matchedSteps,
          },
          { sessionId },
        );
      }

      messages.push(...orderWarnings);
      result.systemMessage = `[flow-progress-tracker] ${messages.join(" | ")}`;
    }
  } catch (error) {
    // Hook execution failure should not block Claude Code
    await logHookExecution(
      HOOK_NAME,
      "approve",
      "Hook execution error (non-blocking)",
      { error: String(error) },
      { sessionId },
    );
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
